src/storage/storages.rs

```diff
@@ -154,112 +178,207 @@ impl<T> SliceAccess<T> for DenseVecStorage<T> {
+        // `Index::MAX` if there is still an entity without this component.
+        unsafe { self.data_id.get_unchecked_mut(id) }.write(self.data.len() as Index);
+        // NOTE: `id` originally of the type `Index` so the cast back won't
+        // overflow.
         self.entity_id.push(id as Index);
-        self.data.push(v);
+        self.data.push(SyncUnsafeCell::new(v));
     }
 
     unsafe fn remove(&mut self, id: Index) -> T {
-        let did = self.data_id.get_unchecked(id as usize).assume_init();
+        // NOTE: cast to usize won't overflow since `insert` would have failed
+        // to allocate.
+        // SAFETY (get_unchecked and assume_init): Caller required to have
+        // called `insert` with this `id`.
+        let did = unsafe { self.data_id.get_unchecked(id as usize).assume_init() };
         let last = *self.entity_id.last().unwrap();
-        self.data_id
-            .get_unchecked_mut(last as usize)
-            .as_mut_ptr()
-            .write(did);
+        // NOTE: cast to usize won't overflow since `insert` would have failed
+        // to allocate.
+        // SAFETY: indices in `self.entity_id` correspond to components present
+        // in this storage so this will be in-bounds.
+        unsafe { self.data_id.get_unchecked_mut(last as usize) }.write(did);
+        // NOTE: casting the index in the dense data array to usize won't
+        // overflow since the maximum number of components is limited to
+        // `Index::MAX + 1`.
         self.entity_id.swap_remove(did as usize);
-        self.data.swap_remove(did as usize)
+        self.data.swap_remove(did as usize).0.into_inner()
+    }
+}
+
+impl<T> SharedGetMutStorage<T> for DenseVecStorage<T> {
+    unsafe fn shared_get_mut(&self, id: Index) -> &mut T {
+        // NOTE: `as` cast is not lossy since insert would have encountered an
+        // allocation failure if this would overflow `usize.`
+        // SAFETY (get_unchecked and assume_init): Caller required to call
+        // `insert` with this `id` (with no following call to `remove` with that
+        // id or to `clean`).
+        let did = unsafe { self.data_id.get_unchecked(id as usize).assume_init() };
+        // SAFETY: Indices retrieved from `data_id` with a valid `id` will
+        // always correspond to an element in `data`.
+        let ptr = unsafe { self.data.get_unchecked(did as usize) }.get();
+        // SAFETY: See `VecStorage` impl.
+        unsafe { &mut *ptr }
     }
 }
 
+// SAFETY: `shared_get_mut` doesn't perform any overlapping mutable
+// accesses when provided distinct indices and is safe to call from multiple
+// threads at once.
 unsafe impl<T> DistinctStorage for DenseVecStorage<T> {}
 
 /// A null storage type, used for cases where the component
 /// doesn't contain any data and instead works as a simple flag.
-pub struct NullStorage<T>(T);
+pub struct NullStorage<T>(PhantomData<T>);
 
-impl<T> UnprotectedStorage<T> for NullStorage<T>
-where
-    T: Default,
```

Comment:
NullStorage no longer requires `T: Default`, since we can soundly conjure up ZSTs provided that they represent instances that this has previously taken ownership of